README.md
# secp256k1Cipher
Encrypt and decrypt data use secp256k1


CHANGELOG.md
Changelog
=========


#### Version 0.1.0 (2019-06-06)

* Add example code and this file.

#### Version 0.1.1 (2019-06-10)

* 1. sha256 compress the secret length to 32, just fix the salsa20 key langth;
* 2. Change IV to base64 string, will be shorter for transport;

### Version 0.1.2 (2019-06-21)
* 1. add uint8list encrypt decrypt support

### Version 0.1.3 (2019-07-02)
* 1. add strict check for point on curve
* 2. fix bug's on serialization of public key

### Version 0.1.4 (2019-07-03)
* 1. add compressed publickey support

### Version 0.1.5 (2019-07-03)
* 1. Change secret to encrypt key method to eliminate ambiguity;

### Version 0.1.6 (2019-07-11)
* 1. fix bugs when encrypt and decrypt unicode String

### Version 0.1.7 (2019-07-11)
* 1. support use part of secret as iv

### Version 0.1.8 (2019-08-03)
* 1. fix bugs when generate iv

### Version 0.2.0 (2019-08-15)
* 1. PublicKey generate BTC Address
* 2. PublicKey generate ETH Address

### Version 0.2.1 (2019-09-12)
* 1. Remove unused logs.

### Version 0.2.2 (2020-10-31)
* 1. Add private key signing and public key verification.

test/impl_test.dart
library secp256k1cipher.test.impl_test;

import 'dart:convert' as convert;
import 'dart:io';
import 'dart:typed_data';
import 'package:secp256k1cipher/src/secp256k1Cipher.dart';
import "package:test/test.dart";
import 'package:secp256k1cipher/secp256k1cipher.dart';
import "package:pointycastle/ecc/api.dart";
import 'package:pointycastle/digests/ripemd160.dart';
import "package:pointycastle/pointycastle.dart";
import "package:hex/hex.dart";
import 'package:base58check/base58.dart';

String _formatBytesAsHexString(Uint8List bytes) {
  var result = StringBuffer();
  for (var i = 0; i < bytes.lengthInBytes; i++) {
    var part = bytes[i];
    result.write('${part < 16 ? '0' : ''}${part.toRadixString(16)}');
  }
  return result.toString();
}

void main() {
  group('Keys', () {
    test("genaddr", () {
      Digest sha256 = new Digest("SHA-256");
      Digest ripemd = new RIPEMD160Digest();
      final pubkey = loadPublicKey(
          '50863AD64A87AE8A2FE83C1AF1A8403CB53F53E486D8511DAD8A04887E5B23522CD470243453A299FA9E77237716103ABC11A1DF38855ED6F2EE187E9C582BA6');
      final pub_bytes = pubkey.Q.getEncoded(false);
      final sha_hash = sha256.process(pub_bytes);
      final rip_hash = ripemd.process(sha_hash);
      final hex_hash = HEX.encode(rip_hash.toList());
      print(hex_hash);

      // 生成验证
      final network_hash = [0x00] + rip_hash.toList();
      final check1 = sha256.process(Uint8List.fromList(network_hash));
      final check2 = sha256.process(check1);
      final final_check = check2.sublist(0, 4);
      final code_list = network_hash + final_check;
      print(HEX.encode(code_list));
      Base58Encoder b58 = new Base58Encoder(
          '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz');
      final b58_str = b58.convert(code_list);
      print(b58_str);
    });
    test('Generate Keys', () {
      File f = new File('/Users/alex/study/test.csv');
      final lines = [];
      for (var _idx = 0; _idx < 100; _idx++) {
        final keypare = generateKeyPair();
        final ECPublicKey pubkey = keypare.publicKey;
        final ECPrivateKey prikey = keypare.privateKey;
        final line = [
          strinifyPublicKey(pubkey),
          strinifyPrivateKey(prikey),
          pubkey.Q.x.toBigInteger().toRadixString(16),
          pubkey.Q.y.toBigInteger().toRadixString(16)
        ];
        final row = line.join(',');
        lines.add(row);
      }
      final txt = lines.join('\n');
      f.writeAsStringSync(txt);
      expect(true, equals(true));
    });
    test('Save and restore private key', () {
      final local_private =
          'eaa692953a60ff85beecdf9647807f5e1bd665aa342c3c1d893b54bccf816ff5';
      final remote_public =
          '02766171786852c788bfac4622b302b1c42ca77e3bfdabc56454a4ca5647ac4eba';
      final enc = 'd5dTsgku25ylogZ7Yjs=';
      final iv = 'MRz0cLx8QL4=';
      final raw = privateDecrypt(local_private, remote_public, enc, iv);
      print('raw: ${raw}');
    });
    test('Save and restore public key', () async {
      for (var _idx = 0; _idx < 100; _idx++) {
        final alic_pubkey =
            '3d6b2142489ffa6d221da41e75e6c08a44a8d0e682b9fa6d768594d94da2adeeb85e248fa05dedcc4f95c32ab8707bb0ba579fd4b41bf28a0df5bfd7f731b809';
        final alic_prikey =
            '462c58255c68a0a1c1b5c89baa99688c81760169ed8c1502d53e50a820aed90a';
        final bob = generateKeyPair();
        print('success idx: ${_idx}');
        final s1 = rawSecret(alic_prikey, strinifyPublicKey(bob.publicKey));
        final s2 = rawSecret(strinifyPrivateKey(bob.privateKey), alic_pubkey);
        expect(s1, equals(s2));
      }
    });
    test('Make request', () async {
      final convert.Utf8Encoder encoder = new convert.Utf8Encoder();
      final remote_pubkey =
          '02be8d8a7b5056de7a7074236100d094ebe86cce33d62469956203022af1f3e556';
      final my_kp = generateKeyPair();
      final data = 'abcdefg';
      final str_pri_key = strinifyPrivateKey(my_kp.privateKey);
      final str_pub_key = strinifyPublicKey(my_kp.publicKey);
      final enced = pubkeyEncryptRaw(str_pri_key, remote_pubkey,
          new Uint8List.fromList(encoder.convert(data)));
      final data_arr = new List<int>();
      data_arr.addAll(encoder.convert(str_pub_key));
      data_arr.addAll(enced['enc'].toList());
      print(convert.base64.encode(data_arr));
      expect(true, true);
    });
    test('Test Decrypt', () async {
      final convert.Utf8Decoder decoder = new convert.Utf8Decoder();
      final my_private =
          '1241ae561074f703c259da27036af3510640bbd6a79ceed7eaea4b3b566befe9';
      final message =
          'MDNjMThhN2RlN2I3ZjQwYTgwMDQwMDg1OGUyMTIwNmYyNzdiYjJhZGMwZjAyMDUzYjMzODYyZDgwY2Q0M2YxN2JhqDMwhGPjj2d4hpz2hfjjyRHQ';
      final raw_data = convert.base64.decode(message);
      final pub_key = decoder.convert(raw_data.getRange(0, 66).toList());
      final payload = raw_data.getRange(66, raw_data.length).toList();
      final decrypted = privateDecryptRaw(
          my_private, pub_key, new Uint8List.fromList(payload));
      print("raw message=${decoder.convert(decrypted)}");
      expect(true, true);
    });
    test('Sign and Verify', () {
      final alice = generateKeyPair();
      final message = "Mary has a little sheep";
      final alice_pubkey = strinifyPublicKey(alice.publicKey);
      final alice_privatekey = strinifyPrivateKey(alice.privateKey);
      final signature = privateSign(alice_privatekey, message);
      expect(true, publicVerify(alice_pubkey, message, signature));
    });
    test('Encrypt and Decrypt', () {
      int micro_seconds = 0;
      for (var i = 0; i < 10; i++) {
        final alic_pubkey =
            '5cb38e0c76f2b28e112e78d96d46e79b04585f17c3bb81a11ad3ad327d9ccaf815b0d2c770fd31c7224671378d7129cdd3dba97ca1efd016e2a580048c6eec46';
        final alic_prikey =
            '9717f155a64b67e5aa22a9552824237119a373b84ffe62eb435cac6581099767';
        var bob = generateKeyPair();
        var raw_str = '测试测试测试';
        final t1 = new DateTime.now().millisecondsSinceEpoch;
        var enc_map = pubkeyEncrypt(
            alic_prikey, strinifyPublicKey(bob.publicKey), raw_str);
        micro_seconds += (new DateTime.now().millisecondsSinceEpoch - t1);
        var enc_str = enc_map['enc'];
        var iv = enc_map['iv'];
        var decryptd = privateDecrypt(
            strinifyPrivateKey(bob.privateKey), alic_pubkey, enc_str);
        print('d:${decryptd}');
        expect(raw_str, equals(decryptd));
      }
      print('avg: ${micro_seconds / 100} ms');
    });
  });
}


example/main.dart
import 'package:secp256k1cipher/secp256k1cipher.dart';

void main(){
  var alic = generateKeyPair(); // Create Alic keypair
  var bob = generateKeyPair();  // Create Bob keypair
  var raw_str = 'Encrypt and decrypt data use secp256k1';  // This is what alic want to say to bob
  var enc_map = pubkeyEncrypt(strinifyPrivateKey(alic.privateKey), strinifyPublicKey(bob.publicKey), raw_str); // use alic's privatekey and bob's publickey means alic say to bob
  var enc_str = enc_map['enc']; // Get encrypted base64 string
  var iv = enc_map['iv'];       // Get random IV
  // next thing, you can send enc_str and IV via internet to bob
  var decryptd = privateDecrypt(strinifyPrivateKey(bob.privateKey), strinifyPublicKey(alic.publicKey), enc_str, iv); // use bob's privatekey and alic's publickey means bob can read message from alic
  print('alic say:${decryptd}');
}

lib/src/operator.dart

import 'package:pointycastle/export.dart';

BigInt theP = BigInt.parse("fffffffffffffffffffffffffffffffffffffffffffffffffffffffefffffc2f", radix: 16);
BigInt theN = BigInt.parse(
          "fffffffffffffffffffffffffffffffebaaedce6af48a03bbfd25e8cd0364141",
          radix: 16);

bool isOnCurve(ECPoint point){
  final x = point.x.toBigInteger();
  final y = point.y.toBigInteger();
  final rs = (y * y - x * x * x - point.curve.a.toBigInteger() * x - point.curve.b.toBigInteger()) % theP;
  return rs == BigInt.from(0);
}

BigInt inverse_mod(BigInt k, BigInt p){
  if (k == 0){
    throw IntegerDivisionByZeroException();
  }
  if (k < BigInt.from(0)){
    return p - inverse_mod(-k, p);
  }
  var s = [BigInt.from(0), BigInt.from(1), BigInt.from(1)];
  var t = [BigInt.from(1), BigInt.from(0), BigInt.from(0)];
  var r = [p, k, k];
  while(r[0]!=BigInt.from(0)){
    var quotient = r[2] ~/ r[0];
    r[1] = r[2] - quotient * r[0];
    r[2] = r[0];
    r[0] = r[1];
    s[1] = s[2] - quotient * s[0];
    s[2] = s[0];
    s[0] = s[1];
    t[1] = t[2] - quotient * t[0];
    t[2] = t[0];
    t[0] = t[1];
  }
  final gcd = r[2];
  final x = s[2];
  final y = t[2];
  assert(gcd == BigInt.from(1));
  assert((k * x) % p == BigInt.from(1));
  return x % p;
}

ECPoint point_neg(ECPoint point){
  assert(isOnCurve(point));
  if (point == null){
    return null;
  }
  final x = point.x.toBigInteger();
  final y = point.y.toBigInteger();
  final result = point.curve.createPoint(x, -y % theP);
  assert(isOnCurve(result));
  return result;
}

ECPoint point_add(ECPoint point1, ECPoint point2){
  if (point1 == null){
    return point2;
  }
  if (point2 == null){
    return point1;
  }
  assert(isOnCurve(point1));
  assert(isOnCurve(point2));
  final x1 = point1.x.toBigInteger();
  final y1 = point1.y.toBigInteger();
  final x2 = point2.x.toBigInteger();
  final y2 = point2.y.toBigInteger();
  if (x1 == x2 && y1 != y2){
    return null; 
  }
  var m;
  if (x1 == x2){
    m = (BigInt.from(3) * x1 * x1 + point1.curve.a.toBigInteger()) * inverse_mod(BigInt.from(2) * y1, theP);
  }else{
    m = (y1 - y2) * inverse_mod(x1 - x2, theP);
  }
  final x3 = m * m - x1 -x2;
  final y3 = y1 + m * (x3 - x1);
  ECPoint result = point1.curve.createPoint(x3 % theP, -y3 % theP);
  assert(isOnCurve(result));
  return result;
}

ECPoint scalar_multiple(BigInt k, ECPoint point){
  assert(isOnCurve(point));
  if (k % theN == 0 || point == null){
    return null;
  }
  if (k < BigInt.from(0)){
    return scalar_multiple(-k, point_neg(point));
  }
  var result = null;
  var addend = point;
  while(k > BigInt.from(0)){
     if (k & BigInt.from(1) > BigInt.from(0)){
       result = point_add(result, addend);
     }
     addend = point_add(addend, addend);
     k >>= 1;
  }
  assert(isOnCurve(result));
  return result;
}



lib/src/secp256k1Cipher.dart
import "dart:typed_data";
import "dart:math";
import 'dart:convert' as convert;
import "package:pointycastle/pointycastle.dart";
import "package:pointycastle/export.dart";
import "package:pointycastle/api.dart";
import "package:pointycastle/ecc/api.dart";
import "package:pointycastle/ecc/curves/secp256k1.dart";
import "package:pointycastle/random/fortuna_random.dart";
import 'package:pointycastle/stream/salsa20.dart';
import 'package:hex/hex.dart';
import 'package:base58check/base58.dart';
import 'package:pointycastle/digests/ripemd160.dart';
import 'operator.dart';
import 'package:pointycastle/src/impl/secure_random_base.dart';
import "package:pointycastle/src/registry/registry.dart";
import "package:pointycastle/src/ufixnum.dart";

class NullSecureRandom extends SecureRandomBase {
  static final FactoryConfig FACTORY_CONFIG =
      new StaticFactoryConfig(SecureRandom, "Null", () => NullSecureRandom());

  var _nextValue = 0;

  String get algorithmName => "Null";

  void seed(CipherParameters params) {}

  int nextUint8() => clip8(_nextValue++);
}

/// return a hex string version privateKey
String strinifyPrivateKey(ECPrivateKey privateKey) {
  return privateKey.d.toRadixString(16);
}

String left_padding(String s, int width) {
  final padding_data = '000000000000000';
  final padding_width = width - s.length;
  if (padding_width < 1) {
    return s;
  }
  return "${padding_data.substring(0, padding_width)}${s}";
}

/// return a BTC Address
String btcAddress(ECPublicKey pubkey) {
  Digest sha256 = new Digest("SHA-256");
  Digest ripemd = new RIPEMD160Digest();
  final pub_bytes = pubkey.Q.getEncoded(false);
  final sha_hash = sha256.process(pub_bytes);
  final rip_hash = ripemd.process(sha_hash);
  // 生成验证
  final network_hash = [0x00] + rip_hash.toList();
  final check1 = sha256.process(Uint8List.fromList(network_hash));
  final check2 = sha256.process(check1);
  final final_check = check2.sublist(0, 4);
  final code_list = network_hash + final_check;
  Base58Encoder b58 = new Base58Encoder(
      '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz');
  return b58.convert(code_list);
}

const int _shaBytes = 256 ~/ 8;
final SHA3Digest sha3digest = SHA3Digest(_shaBytes * 8);

String ethAddress(ECPublicKey pubkey) {
  sha3digest.reset();
  final pub_bytes = pubkey.Q.getEncoded(false);
  final address_bytes = sha3digest.process(pub_bytes);
  final hex_string = HEX.encode(address_bytes);
  return '0x${hex_string.substring(24)}';
}

/// return a hex string version publicKey
String strinifyPublicKey(ECPublicKey publicKey) {
  Uint8List compressedKey = publicKey.Q.getEncoded(true);
  final code_list = compressedKey.toList();
  //print('raw codes:${code_list}');
  return code_list.map((w) {
    final hx = w.toRadixString(16);
    if (hx.length < 2) {
      return '0${hx}';
    }
    return hx;
  }).join('');
  //print('bytes:${raw_bytes}');
  //var x_str = left_padding(publicKey.Q.x.toBigInteger().toRadixString(16), 64);
  //var y_str = left_padding(publicKey.Q.y.toBigInteger().toRadixString(16), 64);
  //return "${x_str}${y_str}";
}

String privateSign(String strPrivateKey, String message) {
  ECPrivateKey privateKey = loadPrivateKey(strPrivateKey);
  ECDSASigner singer = new ECDSASigner(SHA512Digest(), new Mac('SHA-512/HMAC'));
  var privParams = new PrivateKeyParameter(
      new ECPrivateKey(privateKey.d, privateKey.parameters));
  var signParams =
      () => new ParametersWithRandom(privParams, new NullSecureRandom());
  singer.init(true, signParams());
  ECSignature signature = singer
      .generateSignature(Uint8List.fromList(convert.utf8.encode(message)));
  final x_s = signature.r.toRadixString(16);
  final y_s = signature.s.toRadixString(16);
  final hex_x = left_padding(x_s, 64);
  final hex_y = left_padding(y_s, 64);
  return hex_x + hex_y;
}

bool publicVerify(String strPublicKey, String message, String strSignature) {
  ECPublicKey publicKey = loadPublicKey(strPublicKey);
  ECDSASigner verifySinger =
      new ECDSASigner(SHA512Digest(), new Mac('SHA-512/HMAC'));
  var pubkeyParam = new PublicKeyParameter(
      new ECPublicKey(publicKey.Q, publicKey.parameters));

  final str_r = strSignature.substring(0, 64);
  final str_s = strSignature.substring(64, 128);
  final r = BigInt.parse(str_r, radix: 16);
  final s = BigInt.parse(str_s, radix: 16);

  ECSignature signature = new ECSignature(r, s);
  verifySinger.init(false, pubkeyParam);
  return verifySinger.verifySignature(
      Uint8List.fromList(convert.utf8.encode(message)), signature);
}

/// return a privateKey from hex string
ECPrivateKey loadPrivateKey(String storedkey) {
  final d = BigInt.parse(storedkey, radix: 16);
  final param = ECCurve_secp256k1();
  return new ECPrivateKey(d, param);
}

/// return a publicKey from hex string
ECPublicKey loadPublicKey(String storedkey) {
  final param = ECCurve_secp256k1();
  if (storedkey.length < 120) {
    var code_list = new List<int>();
    for (var _idx = 0; _idx < storedkey.length - 1; _idx += 2) {
      final hex_str = storedkey.substring(_idx, _idx + 2);
      code_list.add(int.parse(hex_str, radix: 16));
    }
    final Q = param.curve.decodePoint(code_list);
    return new ECPublicKey(Q, param);
  } else {
    final x = BigInt.parse(storedkey.substring(0, 64), radix: 16);
    final y = BigInt.parse(storedkey.substring(64), radix: 16);
    final Q = param.curve.createPoint(x, y);
    return new ECPublicKey(Q, param);
  }
}

/// return a ECPoint data secret
ECPoint rawSecret(String privateString, String publicString) {
  final privateKey = loadPrivateKey(privateString);
  final publicKey = loadPublicKey(publicString);
  final secret =
      scalar_multiple(privateKey.d, publicKey.Q); //publicKey.Q * privateKey.d;
  //final secret = publicKey.Q * privateKey.d;
  return secret;
}

/// return a Bytes data secret
List<List<int>> byteSecret(String privateString, String publicString) {
  final secret = rawSecret(privateString, publicString);
  final x_s = secret.x.toBigInteger().toRadixString(16);
  final y_s = secret.y.toBigInteger().toRadixString(16);
  final hex_x = left_padding(x_s, 64);
  final hex_y = left_padding(y_s, 64);
  final secret_bytes = Uint8List.fromList(HEX.decode('${hex_x}${hex_y}'));
  final pair = [
    secret_bytes.getRange(0, 32).toList(),
    secret_bytes.getRange(32, 40).toList()
  ];
  //print(secret_bytes);
  //print(pair);
  return pair;
}

/// Encrypt data using target public key
Map pubkeyEncrypt(String privateString, String publicString, String message) {
  convert.Utf8Encoder encoder = new convert.Utf8Encoder();
  final enced = pubkeyEncryptRaw(privateString, publicString,
      Uint8List.fromList(encoder.convert(message)));
  //print('enced:${enced["enc"]}');
  return {'enc': convert.base64.encode(enced['enc']), 'iv': enced['iv']};
}

Map pubkeyEncryptRaw(
    String privateString, String publicString, Uint8List data) {
  final secret_iv = byteSecret(privateString, publicString);
  final secret = Uint8List.fromList(secret_iv[0]);
  final iv = Uint8List.fromList(secret_iv[1]);
  //print('s:${secret} iv:${iv}');
  Salsa20Engine _cipher = Salsa20Engine();
  _cipher.reset();
  _cipher.init(true, _buildParams(secret, Uint8List, iv));
  final Uint8List enc_data = _cipher.process(data);
  return {'enc': enc_data, 'iv': convert.base64.encode(iv)};
}

/// Decrypt data using self private key
String privateDecrypt(
    String privateString, String publicString, String b64encoded,
    [String b64IV = ""]) {
  Uint8List encd_data = convert.base64.decode(b64encoded);
  final raw_data =
      privateDecryptRaw(privateString, publicString, encd_data, b64IV);
  convert.Utf8Decoder decode = new convert.Utf8Decoder();
  return decode.convert(raw_data.toList());
}

Uint8List privateDecryptRaw(
    String privateString, String publicString, Uint8List encd_data,
    [String b64IV = ""]) {
  final secret_iv = byteSecret(privateString, publicString);
  final secret = Uint8List.fromList(secret_iv[0]);
  final iv = b64IV.length > 6
      ? convert.base64.decode(b64IV)
      : Uint8List.fromList(secret_iv[1]);
  Salsa20Engine _cipher = Salsa20Engine();
  _cipher.reset();
  _cipher.init(false, _buildParams(secret, Uint8List, iv));
  return _cipher.process(encd_data);
}

ParametersWithIV<KeyParameter> _buildParams(Uint8List key, Uint8List, iv) {
  return ParametersWithIV<KeyParameter>(KeyParameter(key), iv);
}

/// Generate Keypair
AsymmetricKeyPair<PublicKey, PrivateKey> generateKeyPair() {
  var keyParams = ECCurve_secp256k1();
  var random = FortunaRandom();
  random.seed(KeyParameter(_seed(32)));
  var n = keyParams.n;
  var nBitLength = n.bitLength;
  var d;
  do {
    d = random.nextBigInteger(nBitLength);
  } while (d == BigInt.zero || (d >= n));
  ECPoint Q = scalar_multiple(d, keyParams.G);
  return new AsymmetricKeyPair(
      new ECPublicKey(Q, keyParams), new ECPrivateKey(d, keyParams));
}

Uint8List _seed(length) {
  var random = Random.secure();
  var seed = List<int>.generate(length, (_) => random.nextInt(256));
  return Uint8List.fromList(seed);
}
